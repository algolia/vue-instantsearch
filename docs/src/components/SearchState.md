---
title: SearchState
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/SearchState.md
---

The SearchState widget allows you reach into the current search state, for example for conditional rendering or other rendering that depends on what the current refinement or results are.

<a class="btn btn-static-theme" href="stories/?selectedKind=SearchState">ðŸ•¹ try out live</a>

## Usage

### custom "autocomplete"

SearchState can be used to put a conditional statement around the hits if there's a query.

```html
<div>
  <ais-search-box />
  <ais-search-state>
    <template slot-scope="{ query }">
      <ais-hits v-if="query">
        <h3
          slot="item"
          slot-scope="{ item }"
          :tabindex="0"
          @click="alert(item)"
          @keyup.enter="alert(item)"
        >
          <ais-highlight :hit="item" attribute="title"/>
        </h3>
      </ais-hits>
    </template>
  </ais-search-state>
</div>
```

### No results

SearchState can be used to put a conditional statement around a message to show if there are no results.

```html
<div>
  <ais-search-box />
  <ais-hits />
  <ais-search-state>
    <template slot-scope="{ query, hits }">
      <p v-if="hits.length === 0">
        No results found for the query: <q>{{ query }}</q>
      </p>
    </template>
  </ais-search-state>
</div>
```

## Banner from query rules

If you have a [query rule with userData](https://www.algolia.com/doc/guides/query-rules/query-rules-usage/#return-user-data) in it, you can use SearchState to get this user data as a banner.

```html
<div>
  <ais-search-box />
  <p>type "documentary"</p>
  <ais-search-state>
    <template slot-scope="{ userData }">
      <div>
        <img
          v-for="{ img } in userData"
          :key="img"
          :src="'https://example.com/images/' + img"
        />
      </div>
    </template>
  </ais-search-state>
</div>
```

## Props

This component does not accept any props.

## Slots

Name | Scope | Description
---|---|---
default | `{ "_rawResults", "query", "parsedQuery", "hits", "index", "hitsPerPage", "nbHits", "nbPages", "page", "processingTimeMS", "aroundLatLng", "automaticRadius", "serverUsed", "timeoutCounts", "timeoutHits", "exhaustiveFacetsCount", "exhaustiveNbHits", "userData", "queryID", "disjunctiveFacets", "hierarchicalFacets", "facets" }` | Slot to override the DOM output

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-SearchState` | The root div of the widget
