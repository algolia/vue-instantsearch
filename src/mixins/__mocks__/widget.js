let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
