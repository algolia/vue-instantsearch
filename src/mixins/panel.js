import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
