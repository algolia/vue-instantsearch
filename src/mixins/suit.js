import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
