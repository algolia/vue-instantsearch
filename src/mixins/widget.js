const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
