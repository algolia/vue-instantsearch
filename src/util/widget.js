export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
