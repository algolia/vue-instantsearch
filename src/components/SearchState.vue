<template>
  <div
    :class="suit()"
    v-if="state && state.results"
  >
    <slot v-bind="state.results">
      <p>Use this component to have a different layout based on a certain state.</p>
      <p>Fill in the slot, and get access to the following things on the <code>slot-scope</code>:</p>
      <pre>{{ Object.keys(state.results) }}</pre>
    </slot>
  </div>
</template>

<script>
import { createWidgetMixin } from '../mixins/widget';

const connectStateResults = (renderFn, unmountFn) => (widgetParams = {}) => ({
  init({ instantSearchInstance }) {
    renderFn(
      {
        results: undefined,
        instantSearchInstance,
        widgetParams,
      },
      true
    );
  },

  render({ results, instantSearchInstance }) {
    const resultsCopy = Object.assign({}, results);
    // delete internal state, not to be exposed
    delete resultsCopy._state;
    renderFn(
      {
        results: resultsCopy,
        instantSearchInstance,
        widgetParams,
      },
      false
    );
  },

  dispose() {
    unmountFn();
  },
});

export default {
  mixins: [createWidgetMixin({ connector: connectStateResults })],
  data() {
    return {
      widgetName: 'SearchState',
    };
  },
};
</script>
