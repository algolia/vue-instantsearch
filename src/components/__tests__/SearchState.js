import { mount } from '@vue/test-utils';
import SearchState from '../SearchState.vue';
import { __setState } from '../../mixins/widget';
jest.mock('../../mixins/widget');

it('renders explanation if no slot is used', () => {
  __setState({
    results: {
      query: 'q',
      hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
      page: 1,
    },
  });
  const wrapper = mount(SearchState);
  expect(wrapper.html()).toMatchSnapshot();
});

it("doesn't render if no results", () => {
  __setState({});
  const wrapper = mount(SearchState);
  expect(wrapper.html()).toBeUndefined();
});

it('gives results to default slot', () => {
  const results = {
    query: 'q',
    hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
    page: 1,
  };

  __setState({
    results,
  });

  mount(SearchState, {
    scopedSlots: {
      default: props => expect(props).toEqual(results),
    },
  });
});

it('allows default slot to render whatever they want (truthy query)', () => {
  __setState({
    results: {
      query: 'q',
      hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
      page: 1,
    },
  });

  const wrapper = mount(SearchState, {
    scopedSlots: {
      default: `
      <template slot-scope="{ query }">
        <p v-if="query">
          Query is here
        </p>
        <p v-else>
          There's no query
        </p>
      </template>`,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('allows default slot to render whatever they want (falsy query)', () => {
  __setState({
    results: {
      query: '',
      hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
      page: 1,
    },
  });

  const wrapper = mount(SearchState, {
    scopedSlots: {
      default: `
      <template slot-scope="{ query }">
        <p v-if="query">
          Query is here
        </p>
        <p v-else>
          There's no query
        </p>
      </template>`,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});
