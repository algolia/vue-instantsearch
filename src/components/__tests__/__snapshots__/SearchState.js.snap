// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows default slot to render whatever they want (falsy query) 1`] = `

<div class="ais-SearchState">
  <p>
    There's no query
  </p>
</div>

`;

exports[`allows default slot to render whatever they want (truthy query) 1`] = `

<div class="ais-SearchState">
  <p>
    Query is here
  </p>
</div>

`;

exports[`renders explanation if no slot is used 1`] = `

<div class="ais-SearchState">
  <p>
    Use this component to have a different layout based on a certain state.
  </p>
  <p>
    Fill in the slot, and get access to the following things on the
    <code>
      slot-scope
    </code>
    :
  </p>
  <pre>
    [
  "query",
  "hits",
  "page"
]
  </pre>
</div>

`;
